lib/src/config_fetcher.dart

```diff
@@ -1,82 +1,201 @@
+import 'dart:async';
+import 'package:dio/adapter.dart';
+import 'package:dio/dio.dart';
+
+import 'data_governance.dart';
+import 'configcat_options.dart';
+import 'mixins.dart';
+import 'json/config.dart';
+import 'json/config_json_cache.dart';
+import 'constants.dart';
+import 'log/configcat_logger.dart';
+
+enum _Status { fetched, notModified, failure }
+
+class _RedirectMode {
+  static const int noRedirect = 0;
+  static const int shouldRedirect = 1;
+  static const int forceRedirect = 2;
+}
 
 class FetchResponse {
-  late Status status;
-  final String body;
-
-  FetchResponse(int statusCode, this.body) {
-    if (statusCode >=200 && statusCode < 300) {
-      this.status = Status.fetched;
-    } else if (statusCode == 304) {
-      this.status = Status.notModified;
-    } else {
-      this.status = Status.failure;
-    }
-  }
+  final _Status status;
+  final Config? config;
+
+  FetchResponse._(this.status, this.config);
 
   bool get isFetched {
-    return status == Status.fetched;
+    return status == _Status.fetched;
   }
 
   bool get isNotModified {
-    return status == Status.notModified;
+    return status == _Status.notModified;
   }
 
   bool get isFailed {
-    return status == Status.failure;
+    return status == _Status.failure;
+  }
+
+  factory FetchResponse.success(Config config) {
+    return FetchResponse._(_Status.fetched, config);
+  }
+
+  factory FetchResponse.failure() {
+    return FetchResponse._(_Status.failure, null);
+  }
+
+  factory FetchResponse.notModified() {
+    return FetchResponse._(_Status.notModified, null);
   }
 }
 
-class ConfigFetcher {
-  static const _version = "7.2.0";
-  final Logger log;
-  late String etag = "";
+abstract class Fetcher {
+  Dio get client;
+
+  Future<FetchResponse> fetchConfiguration();
+
+  void close();
+}
+
+class ConfigFetcher
+    with ContinuousFutureSynchronizer<FetchResponse>
+    implements Fetcher {
+  static const globalBaseUrl = 'https://cdn-global.configcat.com';
+  static const euOnlyBaseUrl = 'https://cdn-eu.configcat.com';
+  static const _userAgentHeaderName = 'X-ConfigCat-UserAgent';
+  static const _ifNoneMatchHeaderName = 'If-None-Match';
+  static const _etagHeaderName = 'Etag';
+  static const _successStatusCodes = [200, 201, 202, 203, 204];
+
+  final ConfigCatLogger logger;
+  final ConfigJsonCache jsonCache;
   final String mode;
   final String sdkKey;
-  late final bool urlIsCustom;
-  late final String url;
-
-  ConfigFetcher({
-    required this.log,
-    required this.sdkKey,
-    required this.mode,
-    required DataGovernance dataGovernance,
-    baseUrl = "",
-  }) {
-    this.urlIsCustom = !baseUrl.isEmpty;
-    this.url = baseUrl.isEmpty ? dataGovernance.url : baseUrl;
+
+  late final bool _urlIsCustom;
+  late final Dio _client;
+  late String _etag = '';
+  late String _url;
+
+  ConfigFetcher(this.logger, this.sdkKey, this.mode, this.jsonCache,
+      ConfigCatOptions options) {
```

Comment:
It seems to me a good idea to me to have labeled parameters here, so it's easier for the caller to know what is expected:
```suggestion
  ConfigFetcher({
    required this.logger, 
    required this.sdkKey, 
    required this.mode, 
    required this.jsonCache
    required ConfigCatOptions options,
  })
```
Wdyt?

Reply:
I like the idea of named parameters, is it a more appropriate way to get initialization arguments in a dart environment? I'm thinking about changing to this pattern everywhere if so.

Reply:
At least on public classes :)

Reply:
In the end, I used this way everywhere where it made sense, also introduced it to client functions like `client.getValue()` as it's nice to see which param does what. `getValue('someKey', '')` isn't so descriptive as `getValue(key: 'some-key', defaultValue: '')`.
Thanks for the suggestion!